### PR TITLE
Move CWalletDB::ReorderTransactions to CWallet

### PR DESCRIPTION
**Pull request by pstratem**
CWalletDB was never the right place for this.

### CODE REVIEW

File: src/wallet/walletdb.h [hunk 1]
```diff
@@ -155,6 +155,7 @@ class CWalletDB : public CDB
 
     /// This writes directly to the database, and will not update the CWallet's cached accounting entries!
     /// Use wallet.AddAccountingEntry instead, to write *and* update its caches.
+    bool WriteAccountingEntry(const uint64_t nAccEntryNum, const CAccountingEntry& acentry);
```
Comment (paveljanik) [inline src/wallet/walletdb.h]: Move this a line down because of the comment?

**Comment by paveljanik** [comment]
test_bitcoin fails here:
```
Running 210 test cases...
unknown location:0: fatal error in "acc_orderupgrade": signal: illegal opcode; address of failing instruction: 0x10988a98a
wallet/test/accounting_tests.cpp:24: last checkpoint

*** 1 failure detected in test suite "Bitcoin Test Suite"
Database handles still open at environment close
Open database handle: unnamed/wallet_test.dat
```

**Comment by laanwj** [comment]
Concept ACK

**Comment by laanwj** [comment]
Regarding the Travis issue, I can reproduce this locally simply by running test_bitcoin:
```
Thread 1 "test_bitcoin" received signal SIGILL, Illegal instruction.
CWallet::ReorderTransactions (this=<optimized out>) at /.../bitcoin/bitcoin/src/wallet/wallet.cpp:681
681         nOrderPosNext = 0;
(gdb) bt
#0  CWallet::ReorderTransactions (this=<optimized out>) at /.../bitcoin/bitcoin/src/wallet/wallet.cpp:681
#1  0x00005555557c5d19 in accounting_tests::GetResults (results=std::map with 0 elements)
    at /.../bitcoin/bitcoin/src/wallet/test/accounting_tests.cpp:24
#2  0x00005555557c1128 in accounting_tests::acc_orderupgrade::test_method (this=<optimized out>)
    at /.../bitcoin/bitcoin/src/wallet/test/accounting_tests.cpp:58
#3  0x00005555557c0cde in accounting_tests::acc_orderupgrade_invoker ()
    at /.../bitcoin/bitcoin/src/wallet/test/accounting_tests.cpp:32
#4  0x00005555555ceb91 in boost::unit_test::ut_detail::invoker<boost::unit_test::ut_detail::unused>::invoke<void (*)()> (this=<optimized out>, 
    f=<optimized out>) at /usr/include/boost/test/utils/callback.hpp:56
#5  boost::unit_test::ut_detail::callback0_impl_t<boost::unit_test::ut_detail::unused, void (*)()>::invoke (this=0x0)
    at /usr/include/boost/test/utils/callback.hpp:89
#6  0x00007ffff70cecb1 in ?? () from /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.58.0
#7  0x00007ffff70ae996 in boost::execution_monitor::catch_signals(boost::unit_test::callback0<int> const&) ()
   from /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.58.0
#8  0x00007ffff70af1b3 in boost::execution_monitor::execute(boost::unit_test::callback0<int> const&) ()
   from /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.58.0
#9  0x00007ffff70cede2 in boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::unit_test::test_case const&) ()
   from /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.58.0
#10 0x00007ffff70b609e in boost::unit_test::framework_impl::visit(boost::unit_test::test_case const&) ()
   from /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.58.0
#11 0x00007ffff70ec4cb in boost::unit_test::traverse_test_tree(boost::unit_test::test_suite const&, boost::unit_test::test_tree_visitor&) ()
   from /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.58.0
#12 0x00007ffff70ec4cb in boost::unit_test::traverse_test_tree(boost::unit_test::test_suite const&, boost::unit_test::test_tree_visitor&) ()
   from /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.58.0
#13 0x00007ffff70b19f6 in boost::unit_test::framework::run(unsigned long, bool) ()
   from /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.58.0
#14 0x00007ffff70cd287 in boost::unit_test::unit_test_main(bool (*)(), int, char**) ()
   from /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.58.0
#15 0x0000555555751e04 in main (argc=1442473840, argv=0x555556201460) at /usr/include/boost/test/unit_test.hpp:59
```
Assembly code:
```
   0x0000555555a18275 <+517>:   jmpq   0x555555a181f0 <CWallet::ReorderTransactions()+384>
=> 0x0000555555a1827a <+522>:   ud2    
   0x0000555555a1827c <+524>:   mov    %rax,%r14
   0x0000555555a1827f <+527>:   mov    0x60(%rsp),%rdi
   0x0000555555a18284 <+532>:   cmp    %rbx,%rdi
```
Looks lke a jump "between instructions" or such. Weird.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -651,8 +651,80 @@ bool CWallet::EncryptWallet(const SecureString& strWalletPassphrase)
...
+    }
+
+    int64_t& nOrderPosNext = nOrderPosNext;
+    nOrderPosNext = 0;
+    std::vector<int64_t> nOrderPosOffsets;
+    for (TxItems::iterator it = txByTime.begin(); it != txByTime.end(); ++it)
+    {
+        CWalletTx *const pwtx = (*it).second.first;
+        CAccountingEntry *const pacentry = (*it).second.second;
+        int64_t& nOrderPos = (pwtx != 0) ? pwtx->nOrderPos : pacentry->nOrderPos;
```
Comment (paveljanik) [inline src/wallet/wallet.cpp]:
```
wallet/wallet.cpp:684:14: warning: declaration shadows a field of 'CWallet' [-Wshadow]
```

Comment (paveljanik) [inline src/wallet/wallet.cpp]: +wallet/wallet.cpp:684:14: warning: declaration shadows a field of 'CWallet' [-Wshadow]
+    int64_t& nOrderPosNext = nOrderPosNext;
+             ^
+./wallet/wallet.h:659:13: note: previous declaration is here
+    int64_t nOrderPosNext;
+            ^
+wallet/wallet.cpp:684:30: warning: reference 'nOrderPosNext' is not yet bound to a value when used within its own initialization [-Wuninitialized]
+    int64_t& nOrderPosNext = nOrderPosNext;
+             ~~~~~~~~~~~~~   ^~~~~~~~~~~~~
+2 warnings generated.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -651,8 +651,80 @@ bool CWallet::EncryptWallet(const SecureString& strWalletPassphrase)
...
+        txByTime.insert(make_pair(wtx->nTimeReceived, TxPair(wtx, (CAccountingEntry*)0)));
+    }
+    list<CAccountingEntry> acentries;
+    walletdb.ListAccountCreditDebit("", acentries);
+    BOOST_FOREACH(CAccountingEntry& entry, acentries)
+    {
+        txByTime.insert(make_pair(entry.nTime, TxPair((CWalletTx*)0, &entry)));
+    }
+
+    int64_t& nOrderPosNext = nOrderPosNext;
```
Comment (paveljanik) [inline src/wallet/wallet.cpp]: What is this?

Comment (laanwj) [inline src/wallet/wallet.cpp]: Yes, I think this is the culprit. It's creating a strange circular reference.

Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: @pstratem Are you still working on this?

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (pstratem) [inline src/wallet/wallet.cpp]: indeed that is the issue

**Pull request merged**